File: pages/cart_page.py
import re

from playwright.sync_api import Page, expect

from pages.base_page import BasePage


class CartPage(BasePage):
    def __init__(self, page: Page):
        super().__init__(page)
        self.cart_heading = page.locator(
            'h1:has-text("Cart"), h1:has-text("Shopping Cart"), [class*="cart-title"]'
        ).first
        self.empty_cart_message = page.locator(
            ':text("empty"), :text("no items"), :text("Your cart")'
        ).first
        self.cart_items = page.locator(
            '[class*="cart-item"], [data-qa-locator*="cart"], .cart-product'
        )

    def navigate_to_cart(self) -> None:
        self.goto("/cart/")
        self.dismiss_overlays()

    def expect_cart_page_loaded(self) -> None:
        self.page.wait_for_url(re.compile(r"cart", re.I), timeout=20_000)
        heading_visible = False
        try:
            heading_visible = self.cart_heading.is_visible(timeout=10_000)
        except Exception:
            pass
        url_matches = "cart" in self.page.url.lower()
        assert heading_visible or url_matches

    def expect_cart_accessible(self) -> None:
        expect(self.page).to_have_url(re.compile(r"cart|daraz\.pk", re.I))

File: pages/cart.page.ts
import { expect } from '@playwright/test';
import { BasePage } from './base.page';

export class CartPage extends BasePage {
  private readonly cartHeading = this.page.locator(
    'h1:has-text("Cart"), h1:has-text("Shopping Cart"), [class*="cart-title"]',
  ).first();
  private readonly emptyCartMessage = this.page.locator(
    ':text("empty"), :text("no items"), :text("Your cart")',
  ).first();
  private readonly cartItems = this.page.locator(
    '[class*="cart-item"], [data-qa-locator*="cart"], .cart-product',
  );

  async navigateToCart() {
    await this.goto('/cart/');
    await this.dismissOverlays();
  }

  async expectCartPageLoaded() {
    await this.page.waitForURL(/cart/i, { timeout: 20_000 });
    const headingVisible = await this.cartHeading.isVisible({ timeout: 10_000 }).catch(() => false);
    const urlMatches = /cart/i.test(this.page.url());
    expect(headingVisible || urlMatches).toBe(true);
  }

  async expectCartAccessible() {
    await expect(this.page).toHaveURL(/cart|daraz\.pk/);
  }
}
